### ISSUE TITLE
Optimizing multiple scaffolds at once?

### ISSUE THREAD

**Issue by lvkd84** [Acellera/acegen-open#57]
Hi,

Thank you for developing such an amazing framework.

In Acegen, is it possible to finetune an prior to generalize to a diverse set of scaffolds? Looking at the provided config files, it seems like only 1 scaffold can be passed to `promptsmiles`. There is a large set of scaffolds that we are interested in optimizing. Intuitively, we want to finetune a model to be a generalized molecule optimizer for our specific scoring function.

Also, is it possible to pass in prompts without attachment point specified? In this case, what is the expected behavior of Acegen?

**Comment by MorganCThomas** [comment]
Hi,

Hm, to generalize to a diverse set of scaffolds is interesting. I think there a few options:

- I think it is possible do with the PromptSMILES framework, simply by sampling the prompts (the scaffolds) randomly, before optimization, to ensure all scaffolds are explored (and learned). I can implement this but not until next week or so. Roughly how large a scaffold set?
- Alternatively, you apply a substructure matching scoring function as part of the reward as well. Such that if the molecules contain any of your specified scaffolds, it gets a reward. If you use MolScore, this is already available.

With regards to no attachment points, technically this is possible by making every available position an attachment point which is equivalent. It's only a bit of a pain to write out all the scaffolds with all the attachment points - I could also write a function to automate this if it's of interest.

**Comment by lvkd84** [comment]
Thank you for your quick response!

Our set of prompts may possibly be thousands or tens of thousands. Match scoring may be not feasible here. Plus, we want the model to generalize to unseen prompts as well.

Sampling a prompt at a time is possible and can be done directly in the training script. However, I'm a little concerned that having a batch (defined via `num_envs` in the config) of the same prompt may lead to unstable generalization. I'd prefer a batch to have a diverse set of prompts. However, a batch of the same prompt may be better for exploring different combinations of attachment points? What do you think?

Having a function to automate all possible attachment points would be greatly appreciated!

**Comment by MorganCThomas** [comment]
Match scoring will be feasible but it may just take longer to run, but this is also true for superstructure generation (i.e., all attachment points)

My instinct was to have a diverse range of scaffolds within a batch, although perhaps it's best to implement both options and investigate both. Note that no particular scaffold will be learned to be preferred by this approach, they will be uniformly sampled (unless I also implement a prioritization scheme by storing scaffold rewards). 

Anyway, I'll work on this in a week. If you want to try score matching in the meantime it's available in MolScore. Any questions or a template config file let me know.

**Comment by lvkd84** [comment]
Great! Thanks for picking up on this task. Looking forward to the implementation. It would be great help for us. I believe it's a useful functionality in general.

**Comment by MorganCThomas** [comment]
Hi, I've made some progress. Although there will likely be some things to optimize. 

First I made some changes to PromptSMILES https://github.com/compsciencelab/PromptSMILES/pull/2:
- Accepts a list of scaffold prompts
- Automated superstructure labelling

So first you need to update promptsmiles to the latest version.
```
pip install promptsmiles -U
```
I've updated an ACEGEN branch accordingly https://github.com/Acellera/acegen-open/pull/58 (my personal fork)

This means you can supply a single scaffold with unspecified attachment points e.g., 
```yaml
promptsmiles: c1ccccc1 # -> will automatically generate c1(*)c(*)c(*)c(*)c(*)c1(*) under the hood
```
Or you can supply a list of scaffolds (and can mix, specified/unspecified):
```yaml
promptsmiles:
  - c1ccccc1
  - c1cnc(*)c1
  - ... # etc.
```

Or if you have many scaffolds you can supply a file path to file containing the smiles:
```yaml
promptsmiles: /home/me/path/scaffolds.smi
```

> Note: Currently this is only implemented for the GRU models. I will extend this to other models ASAP. 

**Additional feature**
I've also implemented a prototype substitution feature marked with an X (i.e., c1ccXc(*)c1) which will allow free sampling at that position. Therefore the network might place a "c" or "n", for example. I've already noticed you ca not place two substitutions positions sequentially (e.g., c1cXXcc1), so I need to fix this.

**Comment by lvkd84** [comment]
Hi Morgan,

Thanks for the update! We will start working with these changes. Please let us know if there is any further updates. We will also let you know if we run into any problem.
